rust/src/ike/logger.rs

```diff
@@ -0,0 +1,287 @@
+/* Copyright (C) 2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::ike::{IKEState, IKETransaction};
+use super::ipsec_parser::IKEV2_FLAG_INITIATOR;
+use crate::ike::parser::{ExchangeType, IsakmpPayloadType, SaAttribute};
+use crate::jsonbuilder::{JsonBuilder, JsonError};
+use std;
+use std::convert::TryFrom;
+
+const LOG_EXTENDED: u32 = 0x01;
+
+fn add_attributes(transform: &Vec<SaAttribute>, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    for attribute in transform {
+        js.set_string(
+            attribute.attribute_type.to_string().as_str(),
+            format!("{}", attribute.attribute_value.to_string()).as_str(),
+        )?;
+
+        if let Some(numeric_value) = attribute.numeric_value {
+            js.set_uint(
+                format!("{}_raw", attribute.attribute_type).as_str(),
+                numeric_value as u64,
+            )?;
+        } else if let Some(hex_value) = &attribute.hex_value {
+            js.set_string(
+                format!("{}_raw", attribute.attribute_type).as_str(),
+                &hex_value,
+            )?;
+        }
+    }
+
+    return Ok(());
+}
+
+fn log_ike(
+    state: &IKEState, tx: &IKETransaction, flags: u32, jb: &mut JsonBuilder,
+) -> Result<(), JsonError> {
+    jb.open_object("ike")?;
+
+    jb.set_uint("version_major", tx.hdr.maj_ver as u64)?;
+    jb.set_uint("version_minor", tx.hdr.min_ver as u64)?;
+    jb.set_string("init_spi", &tx.hdr.spi_initiator)?;
+    jb.set_string("resp_spi", &tx.hdr.spi_responder)?;
+    jb.set_uint("message_id", tx.hdr.msg_id as u64)?;
+
+    if tx.ike_version == 1 {
+        if let Some(exchange_type) = tx.hdr.ikev1_header.exchange_type {
+            jb.set_uint("exchange_type", exchange_type as u64)?;
+
+            if (flags & LOG_EXTENDED) == LOG_EXTENDED {
+                match exchange_type {
+                    1 => jb.set_string(
+                        "exchange_type_verbose",
```

Comment:
More rust idiomatic way to do this to avoid copy/pasting `exchange_type_verbose` ?

Reply:
I guess, one could only determine the value string in the `match` and then call `jb.set_string()` only once afterwards. Will do that in the next PR.

Reply:
Yeah, probably similar to above if `ExchangeType` has the `FromPrimitive` derive added to it as well.

Reply:
Done as well. Thanks for the hint!